runtime/src/accounts_index.rs

```diff
@@ -159,16 +159,13 @@ impl<'a, T: 'a + Clone> AccountsIndex<T> {
         reclaims: &mut SlotList<T>,
     ) -> Option<T> {
         if let Some(lock) = self.account_maps.get(pubkey) {
-            let mut list = &mut lock.1.write().unwrap();
-            // filter out other dirty entries
+            let list = &mut lock.1.write().unwrap();
+            // filter out other dirty entries from the same slot
             reclaims.extend(list.iter().filter(|(f, _)| *f == slot).cloned());
             list.retain(|(f, _)| *f != slot);
 
             lock.0.fetch_add(1, Ordering::Relaxed);
             list.push((slot, account_info));
-            // now, do lazy clean
-            self.purge_older_root_entries(&mut list, reclaims);
```

Comment:
I think this is generally good direction. But, let's surface the consequences for quick consensus.


Is it ok to accept (1) slightly bigger index size, (2) more visible spikes at each `clean_accounts()`, and (3) amortized longer lookup time for every account index (because of larger `list`)?

If so, why `older_roots` has been purged here to begin with? Just because of lack of independent `older_root` cleaning mechanism at the ancient times?

Also, the index size will be more easily controlled by malicious tx pattern.

I think these concerns are non-issue for now and only hypothetical worries.

As a good by-product, with this relaxed attitude to the index size, we pave a way for rather straightforward implmentation for #11161. Or, even `snapshot` commitments (now that capitalization check (#11927 ) is enabled, there is some need for this as well).

Reply:
@ryoqun, yeah I think 1, 2, an 3 are acceptable tradeoffs.

"If so, why older_roots has been purged here to begin with? Just because of lack of independent older_root cleaning mechanism at the ancient times?"

Yeah, I think it was just a byproduct of the initial implementation :smile_cat: